src/mlir/passes.md

```diff
@@ -0,0 +1,207 @@
+   they are not in region though they will never alias with a write (RBAA).
+ * `*a` and `*b` can alias if `a` and `b` have the same readonly type
+   and it may alias with writes (RBAA).
+ * `*a` and `*b` cannot alias if `a` and `b` have the same `iso` type,
+   because they must point to disjoint regions (RBAA).
+ * `*a` and `*b` cannot alias if `a` and `b` have different concrete types
+   or interface types that are cannot both satisfied by the same concrete type
+   (TBAA).
+
+When the result of the analysis is that it *cannot* alias, we're free to do any
+optimisation straight away. For those that *can*, we need further analysis to
+separate them between *must*, *may* and *must-not*.
+
+This is the analysis that tracks provenance (the origin of the references) to
+make sure they come from disjoint memory. These can come from:
+ * local variables, for example via assignments (`var`, `let`), in which case a
+   local analysis is enough.
+ * function arguments, in which case we need to find all callers of the function
+   and separate between aliasing calls and non-aliasing calls. If we only have
+   non-aliasing calls, we can optimise as before. If not, we may choose to create
+   multiple versions of the function and change the original calls to match.
+ * class fields, which could have been modified by a previous step, depending on
+   runtime behaviour, and are therefore very hard to prove either way.
+
+### IR state afterwards
+
+After the alias analysis step, the IR shouldn't change much in structure.
+
+If we found opportunities for optimisation at the MLIR level, then some code may
+have simplified, but not changed in dialect support. For the rest of the alias
+information, we should annotate the operations to carry on TBAA info into the
+LLVM IR and allow LLVM to optimise on its own.
+
+## Dynamic vs static calls
+
+Static calls are cheaper than dynamic calls. The former is represented in MLIR
+and LLVM IR natively and translate to direct jumps into code memory. The latter
+has to find the object's vtable address, the method's offset and do an indirect
+call on that address.
+
+We will generate a function for every method, taking the receiver as the
+first argument. Calls to concrete types are static call on the type's
+function directly and therefore need no additional treatment.
+
+For all the call sites where the target is an interface type, we should insert
+indirect (vtable) calls for every class that is ever cast to an interface type,
+containing only the methods that might be called via an interface.
+
+This minimises the size and complexity of the vtables and improve runtime
+performance avoiding unnecessary indirect calls.
+
+### IR state afterwards
+
+At this stage, all static calls should have moved to a direct `std.call` to
+named (and possibly mangled) names that exist as functions in the IR, with the
+appropriate lexical context (module name, etc).
+
+## Virtual table colouring
```

Comment:
For the time being, you can stick with a very simple "each name gets a different offset" and not bother do anything smart. It'll be inefficient, but easy to implement and easy to improve later.

Reply:
this is a document on the intention and listing the things we'll need to do, not a plan to do it all now.